### PR TITLE
Fix #98: Add support for 16-bit floating point numbers

### LINKED ISSUE #98: Support for _Float16
I see that #52 adds support for `_Float{32, 64, 128}` which all seem to work, but there seems to be no support for `_Float16`. Is this something that can be added in?

### PR DESCRIPTION
**Pull request by jcmartin**
This adds support for the following 16-bit floating point numbers:

- Support for half-precision floating point `__fp16`, `_Float16`, and `_Float16x`.
- Support for bfloat16 `__bf16`.

This change does not cause any breaking changes and so the minor version was bumped.

### CODE REVIEW

**Comment by jcmartin** [comment]
@Bodigrim I understand that you are doing maintenance for this package at this point. Would you accept a pull request to fix this issue? This is largely affecting the newest distributions of macOS that modified `math.h` to be dependent on `__fp16`, which makes it something of a high priority in order for `c2hs` to continue to work on that platform.

**Comment by Bodigrim** [comment]
> This change does not cause any breaking changes and so the minor version was bumped.

Adding new constructors to exported data types is a breaking change unfortunately. Not a big deal, just bump the major version please.

----

@jcmartin I am more of caretaker than maintainer, I don't really have any proper domain knowledge. If someone is capable to review, I'll be happy to merge and release. 

@vmchale given your comment in #98, might you be interested to take a look?

**Comment by jcmartin** [comment]
@Bodigrim Fair enough. If it is alright, do you mind if we do two pull requests? One that does not have bfloat16 and one that adds it as well. This way people can benefit from the `__fp16` change without having to wait for `c2hs` to update its dependencies?

**Comment by Bodigrim** [comment]
To be frank, I don't have that much bandwidth. It's easier to revise the latest release of `c2hs` than make an additional release of `language-c`.

**Comment by jcmartin** [comment]
I have bumped the version to 0.10.0.

**Comment by elwright777** [comment]
I tested jcmartin's patch, building a version of c2hs with his language-c-0.9.5.  It successfully parsed the `math.h` header from Apple's Xcode 16.1, which had failed before the patch.  (I also noted this as a comment in #93.)

File: src/Language/C/Parser/Lexer.x [hunk 1]
```diff
@@ -328,13 +332,17 @@ label __label__
...
+-- Tokens: _Alignas _Alignof __alignof alignof __alignof__ __asm asm __asm__ _Atomic auto break _Bool case char __const const __const__ continue _Complex __complex__ default do double else enum extern float for _Generic goto if __inline inline __inline__ int __int128_t long _Noreturn _Nullable __nullable _Nonnull __nonnull register __restrict restrict __restrict__ return short __signed signed __signed__ sizeof static _Static_assert struct switch typedef __typeof typeof __typeof__ __thread _Thread_local __uint128 __uint128_t union unsigned void __volatile volatile __volatile__ while __label__ _Float16 _Float16x _Float32 _Float32x _Float64 _Float64x _Float128 __float128 _Float128x __attribute __attribute__ __extension__ __real __real__ __imag __imag__ __builtin_va_arg __builtin_offsetof __builtin_types_compatible_p __builtin_bit_cast __builtin_convertvector
```
Comment (Bodigrim) [inline src/Language/C/Parser/Lexer.x:335]: Just for my understanding: why does it add `_Float16` and `_Float16x`, but not `__bf16` and `__fp16`?

Comment (jcmartin) [reply src/Language/C/Parser/Lexer.x:335]: A mere oversight. This was the output of the generator in `scripts/GenerateKeywords.hs` before I added `__bf16` and `__fp16`. I can correct this if you like. It has no functional consequence.

Comment (Bodigrim) [reply src/Language/C/Parser/Lexer.x:335]: Let's correct it please.

Comment (jcmartin) [reply src/Language/C/Parser/Lexer.x]: Done

**Comment by Bodigrim** [comment]
I'm trying to get `cabal test` to pass before cutting a release. Apparently baselines are off, probably have been broken for a long time. Any help with it will be appreciated and will speed up `language-c-0.10` release.

**Comment by Bodigrim** [comment]
I have not looked at the testing framework, maybe we can sanitize / filter certain lines to achieve uniform behavior?

I'm also on macOS, one way to run `cabal test` under Linux would be to set up a CI job. 

I'd like to have an evidence that tests pass at least for one platform. If it happens to be easier to achieve for Linux / GCC instead of macOS / clang, that's fine.

**Comment by jcmartin** [comment]
I ran these `cabal test` under linux with `gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0`. Everything is passing.

**Comment by Bodigrim** [comment]
Awesome, published: https://hackage.haskell.org/package/language-c-0.10.0

**Pull request merged**